File: Gauss_Seidel.m
function x = Gauss_Seidel(A)
  %% 实现高斯-赛德尔迭代，已设定参数默认值，可直接键入 Guass_Seidel 运行
  %% A：增广矩阵

  %% 设定参数默认值
  if ~exist('A')
    A = [-4 1 1 1 1; 1 -4 1 1 1; 1 1 -4 1 1; 1 1 1 -4 1];
  end

  [M N] = size(A);

    %% 初始化
  x0 = zeros(M, 1);
  xk = ones(M, 1);

  while max(abs(x0 - xk)) >= 0.00001

    xk = x0;

    for i = 1 : M
      if i == 1
        x0(i) = (1/A(i, i)) * (A(i, end) - A(i,i+1:end-1) * xk(i+1:end));
      elseif i == M
        x0(i) = (1/A(i, i)) * (A(i, end) - A(i, 1:i-1) * x0(1:i-1));
      else
        x0(i) = (1/A(i, i)) * (A(i, end) - A(i, 1:i-1) * x0(1:i-1) - A(i,i+1:end-1) * xk(i+1:end));
      end
    end

  end

  x = x0;

end

File: Gauss_Seidel.py
import numpy as np

def Gauss_Seidel(A=[[-4, 1, 1, 1, 1],
                    [1, -4, 1, 1, 1],
                    [1, 1, -4, 1, 1],
                    [1, 1, 1, -4, 1]]):
    """ 高斯-赛德尔迭代
        返回值：方程组解
    Args:
        A (list): 增广矩阵
    """
    
    A = np.array(A,dtype=float)
    [M, N] = np.shape(A)
    
    x0 = np.zeros((M, 1))
    xk = np.ones((M, 1))
    
    while np.max(np.abs(x0 - xk)) >= 0.0000001:
        
        xk = x0
        for i in range(0, M):
            if i == 1:
                x0[i] = (1/A[i, i]) * (A[i, -1] - np.dot(A[i, i+1:-1], xk[i+1:]))
            elif i == M - 1:
                x0[i] = (1/A[i, i]) * (A[i, -1] - np.dot(A[i, :i], x0[:i])) # np.dot 好像可以自动识别前后向量维度来计算
            else:
                x0[i] = (1/A[i, i]) * (A[i, -1] - np.dot(A[i, :i], x0[:i]) - np.dot(A[i, i+1:-1], xk[i+1:]))
            
    return x0

if __name__ == '__main__':

    A = [[20, 2, 3, 24],[1, 8, 1, 12],[2, -3, 15, 30]]
    print(Gauss_Seidel(A))